Dev console fixes and layout improvements

phillip-kruger opened the pull request:
Signed-off-by:Phillip Kruger <phillip.kruger@gmail.com>
Fix #14216

This PR Contains a few small fixes related to the new non-application path and dev console.

**Dev Console**

- Fixed case where httpRoot is set (did not work)
- Pass httpRoot and frameworkRoot to all templates
- Make config available in templates
- Removed hardcoded `/q` from main.html

**Health UI**

- Fixed case where httpRoot is set
- Fixed wrong ui dependency in health UI (runtime vs deploy)
- Added embeded.html to expose health UI links
- Removed hardcoded `health-ui` and replaced with config lookup

**GraphQL UI**

- Fixed case where httpRoot is set
- Removed hardcoded `/q` from embedded.html
- Removed hardcoded `graphql-ui` and replaced with config lookup

**Swagger UI**

- Removed hardcoded `/q` from embedded.html
- Removed hardcoded `swagger-ui` and replaced with config lookup
- Fixed case where httpRoot is set

gsmet committed:
Support deploying from directory in WebJarUtil

gsmet committed:
DEV Console - Use locally deployed resources

Using externally hosted resources will not work when not connected to
the Internet.

gsmet committed:
DEV Console - Make breadcrumb prettier

gsmet committed:
DEV Console - Improve index page

FroMage commented on extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallRyeGraphQLProcessor.java:
```diff
@@ -454,18 +454,19 @@ void getGraphqlUiFinalDestination(
             String graphQLPath = httpRootPath.adjustPath(graphQLConfig.rootPath);
             String graphQLUiPath = nonApplicationRootPathBuildItem
-                    .adjustPath(httpRootPath.adjustPath(graphQLConfig.ui.rootPath));
+                    .adjustPath(graphQLConfig.ui.rootPath);
```
Is it on purpose that you're moving `httpRootPath.adjustPath` here once in favour of doing it later several times?

phillip-kruger replied:
Yea, because `NotFoundPageDisplayableEndpointBuildItem` does this under the covers, so I need a var without this for that.

FroMage replied:
OK

FroMage commented on extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallRyeGraphQLProcessor.java:
```diff
@@ -483,7 +484,8 @@ void getGraphqlUiFinalDestination(
                                         LINE_FORMAT)
                                 .getBytes(StandardCharsets.UTF_8);
                         content = WebJarUtil
-                                .updateUrl(new String(content, StandardCharsets.UTF_8), graphQLPath, UI_LINE_TO_UPDATE,
```
This changes `graphQLPath` for `graphQLUiPath`, is it on purpose?

phillip-kruger replied:
Yes, was a bug in the prev PR. This is used in "home" (image) click

FroMage commented on extensions/smallrye-health/deployment/src/main/resources/dev-templates/embedded.html:
```diff
@@ -0,0 +1,12 @@
+<a href="{frameworkRootPath}health" class="badge badge-light">
```
I'm strongly opposed to `frameworkRootPath` that includes a `/` at the end, because it makes for very poor and confusing template usage like `{frameworkRootPath}health` where inevitably people will wonder if they need to include a `/` or not.

Can you change it to not include any last `/` and use `{frameworkRootPath}/health` in templates? This would make it consistent with the embedded path template parameter too.

phillip-kruger replied:
So, I initially had it like that, but.... to also make sure the case where httpRoot is / and nonApp is switch off, (so path is /) that we do not need to do 'if /' 
Now in all cases I know that I can just add my page. Cases include /, /httpRoot/, /httpRoot/nonAppRoot/, /nonAppRoot/

Do you think I should just add coded to always remove the last / ?

phillip-kruger replied:
Removed the /

FroMage commented on extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/devmode/console/DevConsoleProcessor.java:
```diff
@@ -155,7 +157,15 @@ public void handle(HttpServerRequest event) {
+        // "/" or "/myroot/" 
+        String httpRootPath = httpRootPathBuildItem.adjustPath("/");
+        // "/" or "/myroot/" or "/q/" or "/myroot/q/"
+        String frameworkRootPath = httpRootPathBuildItem.adjustPath(nonApplicationRootPathBuildItem.adjustPath("/"));
```
Those should not have the last `/`.

phillip-kruger replied:
Ok, let me remove that.

phillip-kruger replied:
Removed

FroMage reviewed (changes requested):
Looks good, except two verifications and a disagreement about slashes ;)

phillip-kruger committed:
Non-application path and Dev console fixes

Signed-off-by: Phillip Kruger <phillip.kruger@gmail.com>

gsmet committed:
Allow usage of the Bootstrap Icons font

FroMage commented on extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/devmode/console/DevConsoleProcessor.java:
```diff
@@ -155,7 +157,17 @@ public void handle(HttpServerRequest event) {
+        String httpRootPath = httpRootPathBuildItem.adjustPath("/");
+        httpRootPath = httpRootPath.substring(0, httpRootPath.lastIndexOf("/"));
+        // "" or "/myroot" or "/q" or "/myroot/q"
+        String frameworkRootPath = httpRootPathBuildItem.adjustPath(nonApplicationRootPathBuildItem.adjustPath("/"));
```
Are you absolutely sure that those will have a trailing `/`? There's no way for the user to set them to `/foo`?

phillip-kruger replied:
Yes, adjustPath add that

phillip-kruger replied:
(I tested with /foo)

FroMage replied:
OK great.

gsmet commented:
I just pushed my changes to this branch as we will collaborate with Phillip.

gsmet reviewed (changes requested):
Please don't merge, we are iterating with @phillip-kruger (but I still want the CI to run so don't want to make it draft).

gsmet committed:
Switch to Font Awesome and also pick better icons

Most of the icons were copy/pasted from the first examples.

gsmet commented:
We have a test failure:
```
Error:  io.quarkus.vertx.web.SimpleRouteTest.testSimpleRoute  Time elapsed: 1.125 s  <<< FAILURE!
java.lang.AssertionError: 
1 expectation failed.
Expected status code <405> but was <404>.
```

Looks like the fact that GET on `/delete` is not supported now get you status 404 instead of 405.

kenfinnigan commented on extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/devmode/console/DevConsole.java:
```diff
@@ -35,8 +35,13 @@
+    DevConsole(Engine engine, String httpRootPath, String frameworkRootPath) {
         this.engine = engine;
+        this.globalData.put("httpRootPath", httpRootPath);
+        this.globalData.put("frameworkRootPath", frameworkRootPath);
```
Should this be named "nonApplicationRootPath" instead of "frameworkRootPath" to align with the name elsewhere?

gsmet committed:
DEV Console - Move static resources outside of META-INF/resources/

Otherwise it causes the installation of the root static handler and
Vert.x Web tests start to fail.

gsmet committed:
Allow using a fluid layout when we have large tables

gsmet committed:
DEV Console - Improve layout of forms

phillip-kruger committed:
Added dark top banner and application name
Signed-off-by:Phillip Kruger <phillip.kruger@gmail.com>

mkouba committed:
Dev console - make it possible to output config properties

gsmet committed:
DEV Console - Minor adjustments

mkouba committed:
Dev console config editor - fix NPE if application.properties missing

gsmet committed:
DEV Console - Adjust titles so that they are not redundant

gsmet committed:
DEV Console - Small adjustments to new Qute additions

gsmet committed:
DEV Console - Use Font Awesome icons for GraphQL

phillip-kruger committed:
Replaced the hardcoded paths in the extensions embedded.html and replace with config lookup
Signed-off-by:Phillip Kruger <phillip.kruger@gmail.com>

gsmet committed:
quarkus.application.name and quarkus.application.version are not always set

Especially when running tests.
